docs/jitserver_tuning.md

```diff
@@ -0,0 +1,125 @@
+
+As a rule of thumb, you should have 10-20 JVMs simultaneously connected to a server with 1-2 GB of memory. With respect to CPU resources, in Kubernetes you might want to set a low “request” value at the server (1-2 vCPUs) and a larger “limit” value (4-8 vCPUs) in order to soak all those idle cycles. It is possible to connect even more clients to one server instance if memory and CPU resources are increased, but in general, two medium-sized server instances placed on different nodes are better than a single larger server.
+
+## Alleviating CPU congestion at the server
+
+When too many clients connect to the server, the server can become flooded with compilation requests, leading to increased compilation times and slower start-up/ramp-up for the application. It should be noted that a client JVM issues most of its compilation requests during the start-up phase and ramp-up phase of an application, when load is first applied to it. Thus, from the CPU consumption point of view what matters is the number of clients that start-up or ramp-up concurrently. To alleviate the CPU strain on the server, you can start the client JVMs in a staggered fashion, rather than all at the same time. Sometimes the stagger process happens naturally; for instance, when using Kubernetes horizontal pod autoscaling, additional application instances are launched gradually as the load increases.
+
+Another idea is to use the `-Xjit:enableJITServerHeuristics` command line option at the clients. When this option is present, the client JVMs share some of the compilation burden by performing the cheap compilations locally and send only expensive compilations to the server. What constitutes a cheap compilation is determined by JIT heuristics that look at the method size, optimization level and the amount of CPU and memory available to the JVM. 
+
+## Avoiding memory shortages at the server
+
+Roughly speaking, the server uses two types of memory: 
+1.	“Scratch” memory. This is allocated during a compilation (for JIT internal data structures) and released to the operating system at the end of the compilation.
+2.	“Persistent” memory. This is used for client-session caches and only gets deleted when a client terminates gracefully (or when JITServer purging mechanism is triggered).
+
+The total amount of “scratch” memory at any given moment depends on how many compilations are in progress and how expensive those compilations are. To reduce this amount, you can start the clients in a staggered fashion as suggested above, or reduce the number of compilation threads per client. Note that the latter already happens automatically: when the server senses that it is about to run out of memory, it provides feedback to the connected clients to reduce their number of active compilation threads.

```

Comment:
Please use normal double-quotes (instead of UTF8 'e2 80 9c' or 'e2 80 9d'). Once defined, quotes are not needed around those terms.

Reply:
There are still a number of places that don't use normal quotes.

Reply:
I missed "requests" and "limits". Fixed now.